packages/js/settings-editor/src/tests/route.test.tsx

```diff
@@ -181,7 +181,7 @@ describe( 'route.tsx', () => {
 				.calls[ 0 ][ 2 ];
 			hookAddedCallback( 'unrelated_hook' );
 
-			expect( applyFilters ).toHaveBeenCalledTimes( 1 ); // Only initial call
+			expect( applyFilters ).toHaveBeenCalledTimes( 2 ); // Only initial call
```

Comment:
@chihsuan I'm not entirely sure how to get around this one. `getModernPages()` needs to be called on param change as well as initially so that extensions have a chance to re-render on  param change. See implementation for extensions:

```js
addFilter( 'woocommerce_admin_settings_pages', 'woocommerce', ( pages ) => {
	const currentArgs = getQueryArgs( window.location.href );
	const section = currentArgs.section ? currentArgs.section : 'mammals';

	pages[ 'settings-tester-modern-screen' ] = {
		areas: {
			content: <ModernScreen section={ section } />,
			edit: null,
		},
		widths: {
			content: undefined,
			edit: 380,
		},
	};
	return pages;
} );
```

Reply:
Hey @psealock I understand the issue. We want to update routes when params change, but currently, getModernPages() is being called both in the initial state and in the effect.

I think we can fix it by using `useRef` or [add a top-level variable](https://react.dev/learn/you-might-not-need-an-effect#initializing-the-application):

```ts
export function useModernRoutes(): Record< string, Route > {
    const [ routes, setRoutes ] = useState<Record<string, Route>>(getModernPages());
    const location = useLocation() as Location;
    const isFirstRender = useRef(true);

    // ... existing hookAdded effect ...

    // Only update on param changes after first render
    useEffect(() => {
        if (isFirstRender.current) {
            isFirstRender.current = false;
            return;
        }
        setRoutes(getModernPages());
    }, [location.params]);

    return routes;
}
```

This way we get:

- Immediate routes on first render 
- No double call of getModernPages()
- Updates on param changes after the initial render

Reply:
Oh just seeing this comment now @chihsuan, thanks for the PR on this one.